### ISSUE TITLE
Avoid importing GUI specific functions/files in headless mode

### ISSUE THREAD

**Issue by lucyleeow** [napari/napari#6469]
## 🧰 Task

Although I think currently we have separated files/functions that directly import Qt, there are functions that do not make sense headless (e.g., `_viewer_key_bindings`) that are still being imported in headless mode. Also files that do not make sense without a GUI (e.g., `napari/components/_viewer_mouse_bindings.py`) live with the non-GUI files.

This was noticed because in https://github.com/napari/napari/pull/4991#issuecomment-1811769245 in [headless CI's](https://github.com/napari/napari/actions/runs/6845972380/job/18612201921?pr=4991#step:11:178) the following import trace caused an import error:

```
  napari/conftest.py:61: in <module>
      from napari.components import LayerList
  .tox/py39-linux-headless-no_cov/lib/python3.9/site-packages/_pytest/assertion/rewrite.py:186: in exec_module
      exec(co, module.__dict__)
  napari/components/__init__.py:24: in <module>
      from napari.components import _viewer_key_bindings  # isort:skip
  .tox/py39-linux-headless-no_cov/lib/python3.9/site-packages/_pytest/assertion/rewrite.py:186: in exec_module
      exec(co, module.__dict__)
  napari/components/_viewer_key_bindings.py:5: in <module>
      from napari.components.viewer_model import ViewerModel
  .tox/py39-linux-headless-no_cov/lib/python3.9/site-packages/_pytest/assertion/rewrite.py:186: in exec_module
      exec(co, module.__dict__)
  napari/components/viewer_model.py:66: in <module>
      from napari.plugins.utils import get_potential_readers, get_preferred_reader
  .tox/py39-linux-headless-no_cov/lib/python3.9/site-packages/_pytest/assertion/rewrite.py:186: in exec_module
      exec(co, module.__dict__)
  napari/plugins/__init__.py:7: in <module>
      from napari.plugins import _npe2
  .tox/py39-linux-headless-no_cov/lib/python3.9/site-packages/_pytest/assertion/rewrite.py:186: in exec_module
      exec(co, module.__dict__)
  napari/plugins/_npe2.py:27: in <module>
      from qtpy.QtWidgets import QWidget
```

Should we be stricter in separating GUI related functions?

cc @DragaDoncila

**Comment by Czaki** [comment]
> Also files that do not make sense without a GUI (e.g., `napari/components/_viewer_mouse_bindings.py`) live with the non-GUI files.

These files do not have sense without GUI, but have sense with a different backend than Qt. For example, Jupyter one that is on our target list. 

But may some thing need to be moved

**Comment by lucyleeow** [comment]
> These files do not have sense without GUI, but have sense with a different backend than Qt. 

Yes good point. Maybe the aim is to avoid importing files/functions not required in headless mode, which I think is reasonable. It may be difficult to achieve though. Potentially we could have 3 part separation:

* common files
* GUI specific
* Qt specific
 
I don't know the code base enough, how achievable is this?

**Comment by lucyleeow** [comment]
> Jupyter one that is on our target list.

Maybe such separation would also aid file structure/clean-ness if/when we were to add another backend?

**Comment by Czaki** [comment]
We have import-linter configuration 

https://github.com/napari/napari/blob/596d74d00beff2b9ca5ed7bdc07c1e987dff76ca/setup.cfg#L200-L232

But I do not see when it was removed from CI.

**Comment by lucyleeow** [comment]
>  I agree it would be really good to ensure that there are no Qt imports other than in the base namespace (napari/) and in napari/_qt/. 

I think this is mostly being achieved. This issue is more about GUI specific functions that do not import Qt. E.g., it seems `_viewer_key_bindings` would not be needed in headless mode but it does not directly import Qt so it is being imported in headless mode.

**Comment by lucyleeow** [comment]
> But I do not see when it was removed from CI.

@Czaki I think you did add it to CI in: #2878 and I think it is still there?

**Comment by Czaki** [comment]
No Talley removed it in #3717
And no reviewer noticed that not all steps are executed on pre-commit ci.

**Comment by Czaki** [comment]
Pre commit.ci runs only steps that are executed when create commit. It does not run check that are in manual step. 

And we decide to not run lint check always as the errors may be not readable for all

**Comment by beniroquai** [comment]
Hey, this issues already a little older but I was wondering if there was any work happening related to the headless mode without relying on qt-based functions in general in my ideal world I could use the napari-based plugins without installing qt at all. Would that be possible ever?:)

**Comment by DragaDoncila** [comment]
@beniroquai sorry for the lack of response here! This is definitely still a goal, and aspects of this will be worked on in the coming months, specifically decoupling the viewer core from specific backends like `qt` and `vispy`, and eventually moving towards [napari-lite](https://github.com/napari/napari/issues/5940). Typically when we think of `napari-lite` we don't think of including plugins, as they would be part of the larger viewer machinery. What kind of usage of plugins are you thinking of? Using different functions from the plugins without a viewer e.g. in a Jupyter notebook? Or just having a different backend that's not qt for the viewer interface?
